mapviz_plugins/src/point_drawing_plugin.cpp

```diff
@@ -0,0 +1,485 @@
+      }
+      else
+      {
+        transformed = DrawLaps();
+      }
+    }
+    else if (buffer_size_ == INT_MAX)
+    {
+      buffer_size_ = buffer_holder_;
+      laps_.clear();
+      got_begin_ = false;
+    }
+    if (draw_style_ == ARROWS)
+    {
+      transformed = DrawArrows();
+    }
+    else
+    {
+      transformed = DrawLines();
+    }
+    return transformed;
+  }
+  void PointDrawingPlugin::CollectLaps()
+  {
+    if (!got_begin_)
+    {
+      begin_ = cur_point_.point;
+      points_.clear();
+      buffer_holder_ = buffer_size_;
+      buffer_size_ = INT_MAX;
+      got_begin_ = true;
+    }
+    tf::Point check = begin_ - cur_point_.point;
+    if (((std::fabs(check.x()) <= 3) && (std::fabs(check.y()) <= 3)) &&
+        (new_lap_ == false))
+    {
+      new_lap_ = true;
+      if (points_.size() > 0)
+      {
+        laps_.push_back(points_);
+        laps_[0].pop_back();
+        points_.clear();
+        points_.push_back(cur_point_);
+      }
+    }
+
+    if (((std::fabs(check.x()) > 25) && (std::fabs(check.y()) > 25)) &&
+        (new_lap_ == true))
+    {
+      new_lap_ = false;
+    }
+  }
+
+  bool PointDrawingPlugin::DrawLines()
+  {
+    bool transformed = false;
+    glColor4f(color_.redF(), color_.greenF(), color_.blueF(), 1.0);
+    if (draw_style_ == LINES)
+    {
+      glLineWidth(3);
+      glBegin(GL_LINE_STRIP);
+    }
+    else
+    {
+      glPointSize(6);
+      glBegin(GL_POINTS);
+    }
+
+    std::list<StampedPoint>::iterator it = points_.begin();
+    for (; it != points_.end(); ++it)
+    {
+      if (it->transformed)
+      {
+        glVertex2f(it->transformed_point.getX(), it->transformed_point.getY());
+
+        transformed = true;
+      }
+    }
+
+    if (cur_point_.transformed)
+    {
+      glVertex2f(cur_point_.transformed_point.getX(),
+                 cur_point_.transformed_point.getY());
+
+      transformed = true;
+    }
+
+    glEnd();
+
+    return transformed;
+  }
+
+  bool PointDrawingPlugin::DrawArrows()
+  {
+    bool transformed = false;
+    glLineWidth(2);
+    glBegin(GL_LINES);
+    glColor4f(color_.redF(), color_.greenF(), color_.blueF(), 0.5);
+    std::list<StampedPoint>::iterator it = points_.begin();
+    for (; it != points_.end(); ++it)
+    {
+      if (it->transformed)
+      {
+        glVertex2f(it->transformed_point.getX(), it->transformed_point.getY());
+        glVertex2f(it->transformed_arrow_point.getX(),
+                   it->transformed_arrow_point.getY());
+
+        glVertex2f(it->transformed_arrow_point.getX(),
+                   it->transformed_arrow_point.getY());
+        glVertex2f(it->transformed_arrow_left.getX(),
+                   it->transformed_arrow_left.getY());
+
+        glVertex2f(it->transformed_arrow_point.getX(),
+                   it->transformed_arrow_point.getY());
+        glVertex2f(it->transformed_arrow_right.getX(),
+                   it->transformed_arrow_right.getY());
```

Comment:
I see some copy/pasted blocks of code in here that could be reduced down to a single function.  Maybe turn the block above here into a function that looks something like `void DrawArray(const StampedPoint& point)`...